shared/lib/accounts/accounts.ts

```diff
@@ -35,46 +50,188 @@ export function getUniqueAccountName(
   return candidateName;
 }
 
-///: BEGIN:ONLY_INCLUDE_IF(solana)
-export async function addDiscoveredSolanaAccounts(
-  controllerMessenger: KeyringInternalSnapClientMessenger,
-  entropySource: string,
-  snapKeyring: SnapKeyring,
+export type SnapAccountNameOptions = {
+  chainId?: CaipChainId;
+};
+
+/**
+ * Get the next available Snap account name for a supported non-EVM Snap.
+ *
+ * @param getNextAvailableAccountName - Callback to get the next available account name.
+ * @param snapId - Snap ID.
+ * @param options - Options for this account name.
+ * @param options.chainId - Chain ID for this account if available.
+ * @returns
+ */
+export async function getNextAvailableSnapAccountName(
+  getNextAvailableAccountName: () => Promise<string>,
+  snapId: SUPPORTED_WALLET_SNAP_ID,
+  { chainId }: SnapAccountNameOptions = {},
 ) {
-  const snapId = SOLANA_WALLET_SNAP_ID;
-  const scopes = [SolScope.Mainnet, SolScope.Testnet, SolScope.Devnet];
-  const client = new KeyringInternalSnapClient({
-    messenger: controllerMessenger,
-    snapId,
-  });
-
-  for (let index = 0; ; index++) {
-    const discovered = await client.discoverAccounts(
-      scopes,
-      entropySource,
-      index,
-    );
+  const defaultSnapAccountName = await getNextAvailableAccountName();
+
+  // FIXME: This is a temporary workaround to suggest a different account name for a first party snap.
+  const accountNumber = defaultSnapAccountName.trim().split(' ').pop();
 
-    // We stop discovering accounts if none got discovered for that index.
-    if (discovered.length === 0) {
-      break;
+  switch (snapId) {
+    case BITCOIN_WALLET_SNAP_ID: {
+      if (chainId === MultichainNetworks.BITCOIN_TESTNET) {
+        return `Bitcoin Testnet Account ${accountNumber}`;
+      }
+      return `Bitcoin Account ${accountNumber}`;
+    }
+    case SOLANA_WALLET_SNAP_ID: {
+      // Solana accounts should have in their scope the 3 networks
+      // mainnet, testnet, and devnet. Therefore, we can use this name
+      // for all 3 networks.
+      return `Solana Account ${accountNumber}`;
     }
+    default:
+      return defaultSnapAccountName;
+  }
+}
+
+///: BEGIN:ONLY_INCLUDE_IF(solana,bitcoin)
+export type WalletSnapOptions = {
+  scope?: CaipChainId;
+  derivationPath?: DiscoveredAccount['derivationPath'];
+  entropySource?: string;
+  accountNameSuggestion?: string;
+};
+
+export type WalletSnapClient = {
+  getSnapId(): SnapId;
+
+  createAccount(
+    options: WalletSnapOptions,
+    internalOptions?: SnapKeyringInternalOptions,
+  ): Promise<KeyringAccount>;
+
+  getNextAvailableAccountName(
+    options?: SnapAccountNameOptions,
+  ): Promise<string>;
+};
+
+type MultichainWalletSnapClientMessenger = Messenger<
+  | SnapControllerHandleRequest
+  | AccountsControllerGetNextAvailableAccountNameAction,
+  never
+>;
+
+export class MultichainWalletSnapClient implements WalletSnapClient {
```

Comment:
Question: Maybe this implementation should go into `app/scripts/lib/accounts.ts` instead? Given that we already have an equivalent implementation in `ui/hooks/accounts/useMultichainWalletSnapClient.ts`?